packages/workflow/src/workflow.ts

```diff
@@ -1168,20 +1157,48 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
   const activator = getActivator();
   if (def.type === 'signal') {
     activator.signalHandlers.set(def.name, handler as any);
-    const bufferedSignals = activator.bufferedSignals.get(def.name);
-    if (bufferedSignals !== undefined && handler !== undefined) {
-      activator.bufferedSignals.delete(def.name);
-      for (const signal of bufferedSignals) {
-        activator.signalWorkflow(signal);
-      }
-    }
+    if (handler !== undefined) dispatchBufferedSignals();
   } else if (def.type === 'query') {
     activator.queryHandlers.set(def.name, handler as any);
   } else {
     throw new TypeError(`Invalid definition type: ${(def as any).type}`);
   }
 }
 
+/**
+ * Set a signal handler function that will handle signals calls for non-registered signal names.
+ *
+ * Signals are dispatched to the default signal handler in the order that they were accepted by the server.
+ *
+ * If this function is called multiple times for a given signal or query name the last handler will overwrite any previous calls.
+ *
+ * @param handler a function that will handle signals for non-registered signal names, or `undefined` to unset the handler.
+ */
+export function setDefaultSignalHandler(handler: DefaultSignalHandler | undefined): void {
+  const activator = getActivator();
+  activator.defaultSignalHandler = handler;
+  if (handler !== undefined) dispatchBufferedSignals();
+}
+
+function dispatchBufferedSignals() {
+  const activator = getActivator();
+  const bufferedSignals = activator.bufferedSignals;
+  while (bufferedSignals.length) {
+    if (activator.defaultSignalHandler) {
```

Comment:
While I see why this is correct, it's a bit confusing that we check the default handler first.
Essentially we have 2 distinct functions in one, and I think it'd be clearer if we kept the logic in the signal setter.

Reply:
The difficulty is that handlers might get registered or unregistered from inside a signal handler. The dispatch function therefore must be reentrant-safe.

Imagine a situation where I receive signals (A, A, B, B, C, C), with following handlers:

```
setHandler(signalC, () => {
   setHandler(signalC, undefined);
   setHandler(signalB, () => {
       setHandler(signalB, undefined);
       setDefaultSignalHandler(() => { ... })
   })
})
```

This is why I always reevaluate the whole list, from the beginning, and restart every time a single signal gets dispatched.

Note that the dispatch loop is now O(N^2), but we don't expect N to grow anything big.

A possible optimisation here would be to have a "signal handler changed" counter, and only restart the loop if that counter changed while dispatching a signal. That would bring the dispatch function down to O(N) in most cases, but I don't think that the potential performance gain in extremely rare cases justify the increased complexity that this solution involves.

Reply:
Good point! I didn't account for registration of signal handlers inside signal handlers.
We'll need to check this to ensure ordering.
It's probably worth moving this code to the activator since it's relying on some of its internals.